nucypher/policy/identity.py

```diff
@@ -0,0 +1,402 @@
+            verifying_key=self.verifying_key,
+            encrypting_key=self.encrypting_key,
+            character=self.__character_flag
+        )
+        return payload
+
+    def describe(self, truncate: int = TRUNCATE) -> Dict:
+        description = dict(
+            nickname=self.__nickname,
+            id=self.id.hex(),
+            verifying_key=bytes(self.verifying_key).hex()[:truncate],
+            character=self.character.__name__
+        )
+        if self.character is Bob:
+            description['encrypting_key'] = bytes(self.encrypting_key).hex()[:truncate]
+        return description
+
+    def to_json(self, as_string: bool = True) -> Union[dict, str]:
+        payload = dict(
+            nickname=self.__nickname.decode(),
+            verifying_key=bytes(self.verifying_key).hex(),
+            encrypting_key=bytes(self.encrypting_key).hex(),
+            character=self.character.__name__
+        )
+        if as_string:
+            payload = json.dumps(payload)
+        return payload
+
+    @classmethod
+    def from_character(cls, character: Character, nickname: Optional[str] = None) -> 'Card':
+        flag = getattr(constant_sorrow.constants, character.__class__.__name__.upper())
+        instance = cls(verifying_key=character.public_keys(power_up_class=SigningPower),
+                       encrypting_key=character.public_keys(power_up_class=DecryptingPower),
+                       character_flag=bytes(flag),
+                       nickname=nickname)
+        return instance
+
+    #
+    # Card API
+    #
+
+
+    @property
+    def verifying_key(self) -> UmbralPublicKey:
+        return self.__verifying_key
+
+    @property
+    def encrypting_key(self) -> UmbralPublicKey:
+        return self.__encrypting_key
+
+    @property
+    def id(self) -> HexBytes:
+        return self.__hash(self.__payload)
+
+    @property
+    def nickname(self) -> str:
+        if self.__nickname:
+            return self.__nickname.decode()
+
+    def set_nickname(self, nickname: str) -> None:
+        if len(nickname.encode()) > self.__MAX_NICKNAME_SIZE:
+            raise ValueError(f'New nickname exceeds maximum size ({self.__MAX_NICKNAME_SIZE} bytes)')
+        self.__nickname = nickname.encode()
+
+    @nickname.setter
+    def nickname(self, nickname: str) -> None:
+        self.set_nickname(nickname)
+
+    #
+    # Card Storage API
+    #
+
+    @property
+    def filepath(self) -> Path:
+        identifier = f'{self.nickname}{self.__DELIMITER}{self.id.hex()}' if self.__nickname else self.id.hex()
+        filename = f'{identifier}.{self.__FILE_EXTENSION}'
+        filepath = self.CARD_DIR / filename
+        return filepath
+
+    @property
+    def is_saved(self) -> bool:
+        exists = self.filepath.exists()
+        return exists
+
+    def save(self, encoder: Callable = base64.b64encode, overwrite: bool = False) -> Path:
+        if not self.CARD_DIR.exists():
+            os.mkdir(str(self.card_dir))
```

Comment:
getting the following when the card dir is not present i.e. I deleted it and then tried to create a contact:

```bash
 [82] → nucypher contacts create --type a --verifying-key 027b8c9cf91929e02446709b9d96faca6f6e6377f52eed2f9c14e08ac1db9267e4 --nickname Rando
Traceback (most recent call last):
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/bin/nucypher", line 33, in <module>
    sys.exit(load_entry_point('nucypher', 'console_scripts', 'nucypher')())
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/derek/.local/share/virtualenvs/nucypher-jOttX5sw/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/derek/Documents/Github/repos/forks/kieran/nucypher/nucypher/cli/commands/contacts.py", line 126, in create
    new_card.save()
  File "/Users/derek/Documents/Github/repos/forks/kieran/nucypher/nucypher/policy/identity.py", line 301, in save
    os.mkdir(str(self.card_dir))
AttributeError: 'Card' object has no attribute 'card_dir'
```

Reply:
:heavy_check_mark: - pushed a fix and tests to suit the missing card directory for both the `list` and `Card.save` functions.